💄Improve table of contents styling

olyism opened the pull request:
Small styling tweak for table of contents (TOC) so it aligns better with the overall sidebar look and feel.

## Before

<img width="1436" alt="Screen Shot 2022-06-20 at 9 44 37 am" src="https://user-images.githubusercontent.com/7202667/174514692-72ef50df-2d36-49c5-a4ea-9753079d05e4.png">

- Drop shadow 
- Body font
- No overflow scroll for exceedingly long TOC

## After

Visual tweaks as shown on my local dev:

- Consistent look and feel as per left sidebar nav
- System font
- Position fixed with overflow scrolling

<img width="1400" alt="Screen Shot 2022-06-20 at 12 09 06 pm" src="https://user-images.githubusercontent.com/7202667/174514720-4f1caadd-6ff6-4eef-8c36-d335d48a254d.png">

### Overflow scroll support

<img width="1436" alt="Screen Shot 2022-06-20 at 12 10 58 pm" src="https://user-images.githubusercontent.com/7202667/174514735-f5e215aa-0939-4c27-aec6-6494f264cd67.png">

### Does not affect existing small screen styles

<img width="821" alt="Screen Shot 2022-06-20 at 12 09 18 pm" src="https://user-images.githubusercontent.com/7202667/174514732-dc495f2b-321c-45f3-aea9-15f5d0da0e0d.png">

lachlanhardy commented:
I can't see the lefthand border on the TOC column in Safari, Chrome, or Firefox, so not sure what happened there.
<img width="1147" alt="image" src="https://user-images.githubusercontent.com/3682/174530707-5b509cd9-edaa-4280-8690-0af70ea0ddba.png">
cf https://bk-docs-pr-1600.onrender.com/docs/pipelines/dependencies

Presumably related, but this no longer handles the overlarge table such as: https://bk-docs-pr-1600.onrender.com/docs/pipelines/environment-variables
<img width="1154" alt="image" src="https://user-images.githubusercontent.com/3682/174531001-effedd25-e981-4df8-bb24-83d623f58ebd.png">


Although it turns out, neither does the live site currently. Right hand cut off here is not cropping but the end of the browser window: 
<img width="1146" alt="image" src="https://user-images.githubusercontent.com/3682/174531158-39699fbb-ada9-4488-9c55-827faa41a7a6.png">

lachlanhardy reviewed (commented):
Screenshot is lovely, but as per my comments, I don't see that live. It also fails to handle the overlarge tables on some pages.

olyism commented on app/assets/stylesheets/public/_nav.scss:
```diff
@@ -127,6 +127,7 @@
 
     &--level2 {
       @media (min-width: $screen-md) {
+        background-color: #fff;
```
In case of content overlap, the white background will cover up what's behind the sidebar

olyism commented on app/assets/stylesheets/public/_search.scss:
```diff
@@ -15,6 +15,7 @@
     font-family: inherit;
     font-size: 16px;
     outline: none;
+    margin: 0;
```
Safari form elements have native margins. Force this to 0.

olyism commented on app/assets/stylesheets/public/_toc.scss:
```diff
@@ -2,6 +2,8 @@
 
   @media (min-width: $screen-lg) {
+    background-color: #fff;
+    border-left: #ddd 1px solid;
```
In case of content overlap, the white background will cover up what's behind the sidebar

Fix left border that got accidentally left out.

lachlanhardy commented:
New version looks great!

olyism commented on app/assets/stylesheets/public/_docs.scss:
```diff
@@ -128,8 +128,19 @@ $color-green: rgb(3, 165, 98);
+    word-wrap: break-word;
+
+    code {
+      white-space: normal;
```
Wrap and break words within `th` and `td`. Also wrapping inline `code` so elements like long variables break into a new line and avoids distorting the layout.

lachlanhardy replied:
There were _reasons_ we didn't do this. Mostly around ease of cutting and pasting variable names or keys into scripts and config files. We never satisfactorily resolved it. Perhaps it is time to revisit those ideas, @plaindocs?

olyism commented on app/models/page/renderer.rb:
```diff
@@ -181,11 +181,13 @@ def init_responsive_tables(doc)
-          faux_th = "<th aria-hidden class=\"responsive-table__faux-th\">#{thead_ths[i].children}</th>"
-          
-          td.add_previous_sibling(faux_th)
+      unless thead_ths.empty?
```
Prevent error and skip prepending _faux `th`_ when no `th` are found in responsive tables.

karensawrey commented:
![no-suggestions](https://user-images.githubusercontent.com/27306876/174589231-09df332e-2f87-448f-96a1-77750478852e.png)
Are the search bar suggestions supposed to work in the Render previews?.. If yes, it's a bug. If no, then I'm okay with the rest of the changes in this PR. @olyism

olyism commented:
hi @lachlanhardy thanks for the feedback. I've looked at this again and propose the following changes and commits to address the issue of table distorting page layout.

## Fix table layout and break word to second line

- Use `table-layout: fixed` so width of table isn't decided by the table cells
- Couple the above with word break in `td`, `th` and `code`

The above effect has been applied to all `.Docs__attribute__table` and `.responsive-table`. 

This does mean that long variable names would break into a second line. It's still legible so for now it shouldn't cause developers too much problem.

@karensawrey thoughts on breaking long variable names? ☝️ 

See commit https://github.com/buildkite/docs/pull/1600/commits/33e8907cf2f1e60e2cf14cd9592482cc45169f4e

### Before

<img width="1305" alt="Screen Shot 2022-06-20 at 9 12 47 pm" src="https://user-images.githubusercontent.com/7202667/174589557-94431da1-235e-48b8-8461-553f16d60d78.png">

Table width pushing out the page layout.

### After

<img width="1308" alt="Screen Shot 2022-06-20 at 9 13 37 pm" src="https://user-images.githubusercontent.com/7202667/174589710-18cf836e-088e-418a-8b15-dc4639a7494d.png">

Long words are broken down to second line to preserve the page layout.

## Defensive styling to prevent page layout distortion

I've added CSS tweaks to provide fixed sidebars with solid background color so that, in the event when a page layout gets distorted, it'll gracefully fallback to horizontal scrolling.

I've also added a few Safari tweaks to override some of its browser styles for better cross-browser pixel perfection.

See commit https://github.com/buildkite/docs/pull/1600/commits/1c0af0ad280a9b4461c733b6ed782911edd21a8f

## Future considerations

I think it's worth considering a more holistic approach to address the presentation of table data. There may be ways to better format the tables, use responsive/adaptive technique or even use interactive UI to help developers look up the data they need.

I'll create a backlog ticket to look into this later.

lachlanhardy commented:
Ah, boo. I wrote a comment above, but you've already addressed it more completely. Sorry about that!

Yes, long-standing issue that needs *some* kind of consistent workaround. Maybe this is it!

karensawrey commented:
No worries, @lachlanhardy! The point is - it's time we addressed the issue of long variable names breaking things... Not the first time, not the first iteration on tables (@olyism this is really an ongoing struggle that started waaaaay back!).
